### PR TITLE
fix(logging): fix request logging

### PR DESCRIPTION
**Pull request by udaraweerasinghege**
fixes https://github.com/mozilla/fxa-content-server/issues/5111

### CODE REVIEW

**Comment by vladikoff** [comment]
@udaraweerasinghege this gives us 


> {"Timestamp":1497495137978000000,"Logger":"fxa-content-server","Type":"server.requests.server.requests","Severity":6,"Pid":41792,"EnvVersion":"2.0","Fields":{"msg":"GET /scripts/views/elements/text-input.js 200 0.640"}}

**Comment by vladikoff** [comment]
Here's a good log example from the profile server : 

```
{"Timestamp":1497495339733000000,"Logger":"fxa-profile-server","Type":"server.summary","Severity":6,"Pid":44096,"EnvVersion":"2.0","Fields":{"code":200,"method":"get","errno":0,"path":"/","agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36","t":25,"remoteAddressChain":"[\"127.0.0.1\"]"}}
```

File: server/lib/logging/route_logging.js [hunk 1]
```diff
@@ -30,7 +30,7 @@ module.exports = function () {
           morgan(formats[config.get('route_log_format')], {
             stream: {
               write: function (x) {
-                logger.info(typeof x === 'string' ? x.trim() : x);
+                logger.info('server.requests', typeof x === 'string' ? x.trim() : x);
```
Comment (vladikoff) [inline server/lib/logging/route_logging.js]: See comments

File: server/lib/logging/route_logging.js [hunk 2]
```diff
@@ -13,24 +13,39 @@ const config = require('../configuration');
...
+    'user-agent': req.headers['user-agent']
+  }),
+  'dev-fxa': (tokens, req, res) => JSON.stringify({
+    method: tokens.method(req, res),
+    path: tokens.url(req, res),
+    'response-time': tokens['response-time'](req, res),
+    status: tokens.status(req, res),
+  })
+};
+
```
Comment (vladikoff) [inline server/lib/logging/route_logging.js]: @jrgm @jbuck see log example above ^

**Comment by vladikoff** [comment]
@udaraweerasinghege I'm getting the following error: 

```
fxa-content-server.route.500.ERROR: SyntaxError: Unexpected token .
    at Object.parse (native)
    at Object.config.get.morgan.stream.write (/Users/vladikoff/mozilla/fxa-local-dev/fxa-content-server/server/lib/logging/route_logging.js:48:43)
    at Array.logRequest (/Users/vladikoff/mozilla/fxa-local-dev/fxa-content-server/node_modules/morgan/index.js:130:14)
    at listener (/Users/vladikoff/mozilla/fxa-local-dev/fxa-content-server/node_modules/morgan/node_modules/on-finished/index.js:169:15)
    at onFinish (/Users/vladikoff/mozilla/fxa-local-dev/fxa-content-server/node_modules/morgan/node_modules/on-finished/index.js:100:5)
    at callback (/Users/vladikoff/mozilla/fxa-local-dev/fxa-content-server/node_modules/morgan/node_modules/on-finished/node_modules/ee-first/index.js:55:10)
    at ServerResponse.onevent (/Users/vladikoff/mozilla/fxa-local-dev/fxa-content-server/node_modules/morgan/node_modules/on-finished/node_modules/ee-first/index.js:93:5)
    at emitNone (events.js:72:20)
    at ServerResponse.emit (events.js:166:7)
    at finish (_http_outgoing.js:558:10)
    at nextTickCallbackWith0Args (node.js:436:9)
    at process._tickDomainCallback (node.js:406:13)
Error: Can't set headers after they are sent.
    at ServerResponse.OutgoingMessage.setHeader (_http_outgoing.js:344:11)
    at ServerResponse.header (/Users/vladikoff/mozilla/fxa-local-dev/fxa-content-server/node_modules/express/lib/response.js:730:10)
    at ServerResponse.send (/Users/vladikoff/mozilla/fxa-local-dev/fxa-content-server/node_modules/express/lib/response.js:170:12)
    at done (/Users/vladikoff/mozilla/fxa-local-dev/fxa-content-server/node_modules/express/lib/response.js:967:10)
    at /Users/vladikoff/mozilla/fxa-local-dev/fxa-content-server/node_modules/consolidate/lib/consolidate.js:735:7
    at /Users/vladikoff/mozilla/fxa-local-dev/fxa-content-server/node_modules/consolidate/lib/consolidate.js:144:5
    at Promise._execute (/Users/vladikoff/mozilla/fxa-local-dev/fxa-content-server/node_modules/bluebird/js/release/debuggability.js:300:9)
    at Promise._resolveFromExecutor (/Users/vladikoff/mozilla/fxa-local-dev/fxa-content-server/node_modules/bluebird/js/release/promise.js:483:18)
    at new Promise (/Users/vladikoff/mozilla/fxa-local-dev/fxa-content-server/node_modules/bluebird/js/release/promise.js:79:10)
    at promisify (/Users/vladikoff/mozilla/fxa-local-dev/fxa-content-server/node_modules/consolidate/lib/consolidate.js:137:10)
    at Function.exports.handlebars.render (/Users/vladikoff/mozilla/fxa-local-dev/fxa-content-server/node_modules/consolidate/lib/consolidate.js:725:10)
    at /Users/vladikoff/mozilla/fxa-local-dev/fxa-content-server/node_modules/consolidate/lib/consolidate.js:165:27
    at /Users/vladikoff/mozilla/fxa-local-dev/fxa-content-server/node_modules/consolidate/lib/consolidate.js:99:5
    at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:380:3)
fxa-content-server.route.500.ERROR: SyntaxError: Unexpected token .
    at Object.parse (native)

```

**Comment by vladikoff** [comment]
@udaraweerasinghege got this when loading the page in Chrome via 127.0.0.1:3030

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: eeb8e3ee129ad2a530c04f31064e7f0cbc50f21d
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

**Comment by udaraweerasinghege** [comment]
@vladikoff Had a typo ._. Was testing with `default-fxa` policy. Didn't test `fxa_dev`. They both work now 👍

File: server/lib/logging/route_logging.js [hunk 3]
```diff
@@ -13,24 +13,39 @@ const config = require('../configuration');
  * Enhances connect logger middleware - custom formats.
  * See lib/configuration for usage.
  */
-const formats = {
-  'default_fxa': ':remote-addr - - ":method :url HTTP/:http-version" :status :response-time :res[content-length] ":referrer" ":user-agent"',
-
-  'dev_fxa': ':method :url :status :response-time'
```
Comment (vladikoff) [inline server/lib/logging/route_logging.js]: let's try to use the string one for `dev_fxa`, the new `default_fxa` looks great!

**Comment by vladikoff** [comment]
@udaraweerasinghege let's add a `route_logging.js` test file into the `server/test` directory. Then we can test that both formats are correct. You can modify the `config` (to switch formats) using `proxyquire`.  Example: https://github.com/mozilla/fxa-content-server/blob/878b694b519e90566cc63ce0f2acad1b276ae21b/tests/server/noindex.js#L15

**Comment by vladikoff** [comment]
@udaraweerasinghege to run server tests it is : `npm run test-server`, make sure to restart the server.

File: server/lib/logging/route_logging.js [hunk 4]
```diff
@@ -13,25 +13,45 @@ const config = require('../configuration');
...
+  }),
+  'dev_fxa': (tokens, req, res) => [
+    tokens.method(req, res),
+    tokens.url(req, res),
+    tokens['response-time'](req, res),
+    tokens.status(req, res)
+  ].join(' ')
+};
+
+function loggerHeka(x) {
```
Comment (vladikoff) [inline server/lib/logging/route_logging.js]: I think we got a bit confused with this function here, I thought we would have 2 of them for each type of log format. Seem like we do not need that. Let's just put this logic back right into `write`.

File: server/lib/logging/route_logging.js [hunk 5]
```diff
@@ -13,25 +13,45 @@ const config = require('../configuration');
...
+  if (config.get('route_log_format') === 'dev_fxa') {
+    return logger.info('route', x.trim());
+  }
+  return logger.info('route',  JSON.parse(x));
```
Comment (vladikoff) [inline server/lib/logging/route_logging.js]: double space before `JSON`

File: server/lib/logging/route_logging.js [hunk 6]
```diff
@@ -13,25 +13,45 @@ const config = require('../configuration');
...
 };
 
+const formats = {
+  'default_fxa': (tokens, req, res) => JSON.stringify({
+    'content-length': tokens.res(req, res, 'content-length'),
+    HTTP: req.httpVersion,
+    method: tokens.method(req, res),
+    path: tokens.url(req, res),
+    referrer: req.headers['referer'],
+    'remote-addr': req.ip || req.connection.remoteAddress,
```
Comment (vladikoff) [inline server/lib/logging/route_logging.js]: let's call this `remoteAddressChain`

File: server/lib/logging/route_logging.js [hunk 7]
```diff
@@ -13,25 +13,45 @@ const config = require('../configuration');
...
   next();
 };
 
+const formats = {
+  'default_fxa': (tokens, req, res) => JSON.stringify({
+    'content-length': tokens.res(req, res, 'content-length'),
+    HTTP: req.httpVersion,
+    method: tokens.method(req, res),
+    path: tokens.url(req, res),
+    referrer: req.headers['referer'],
```
Comment (vladikoff) [inline server/lib/logging/route_logging.js]: keep the same `referer` (the original a misspelling of referrer))

File: server/lib/logging/route_logging.js [hunk 8]
```diff
@@ -13,25 +13,45 @@ const config = require('../configuration');
...
 
 // Used when logging is disabled
 const disabled = function (req, res, next) {
   next();
 };
 
+const formats = {
+  'default_fxa': (tokens, req, res) => JSON.stringify({
+    'content-length': tokens.res(req, res, 'content-length'),
+    HTTP: req.httpVersion,
```
Comment (vladikoff) [inline server/lib/logging/route_logging.js]: we can drop `HTTP`

File: server/lib/logging/route_logging.js [hunk 9]
```diff
@@ -13,25 +13,45 @@ const config = require('../configuration');
...
+    method: tokens.method(req, res),
+    path: tokens.url(req, res),
+    referrer: req.headers['referer'],
+    'remote-addr': req.ip || req.connection.remoteAddress,
+    'response-time': tokens['response-time'](req, res),
```
Comment (vladikoff) [inline server/lib/logging/route_logging.js]: let's make this `t`

File: server/lib/logging/route_logging.js [hunk 10]
```diff
@@ -13,25 +13,45 @@ const config = require('../configuration');
...
+    status: tokens.status(req, res),
+    'user-agent': req.headers['user-agent']
```
Comment (vladikoff) [inline server/lib/logging/route_logging.js]: make this `userAgent`

File: server/lib/logging/route_logging.js [hunk 11]
```diff
@@ -13,25 +13,45 @@ const config = require('../configuration');
...
-};
 
 // Used when logging is disabled
 const disabled = function (req, res, next) {
   next();
 };
 
+const formats = {
+  'default_fxa': (tokens, req, res) => JSON.stringify({
+    'content-length': tokens.res(req, res, 'content-length'),
```
Comment (vladikoff) [inline server/lib/logging/route_logging.js]: let's make that `contentLength`

File: server/lib/logging/route_logging.js (hunk 6 above)
Comment (vladikoff) [inline server/lib/logging/route_logging.js]: I filed an issue to make this do a proper addressChain: https://github.com/mozilla/fxa-content-server/issues/5149

We can just keep `remoteAddressChain: req.ip || req.connection.remoteAddress,` in this PR for now and do a follow up.

File: tests/server/logging/route_logging.js [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+define([
+  'intern!object',
+  'intern/chai!assert',
+  'intern/dojo/node!path',
+  'intern/dojo/node!proxyquire',
+  'intern/dojo/node!sinon'
+], function (registerSuite, assert, path, proxyquire, sinon) {
```
Comment (vladikoff) [inline tests/server/logging/route_logging.js]: Awesome work with the tests 👍

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 7fdd596a61573852291d67aa84afea957436eaf1
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md#git-commit-guidelines

---
This message was auto-generated by https://gitcop.com

**Comment by udaraweerasinghege** [comment]
@vladikoff addressed comments

**Comment by vladikoff** [comment]
Thank you @udaraweerasinghege !

**Pull request merged**